api/controllers/ExportController.js

```diff
@@ -0,0 +1,136 @@
+import stringify from 'csv-stringify'
+
+// Toplevel API entrypoint to check auth & route to desired exporter flow based on parameters
+module.exports = {
+  groupData: async function (req, res) {
+    const p = req.allParams()
+
+    if (!p.groupId) {
+      throw new Error("Please specify group ID")
+    }
+    if (!p.datasets || !p.datasets.length) {
+      throw new Error("Please specify datasets to export")
+    }
+
+    // auth check
+    let ok = false
+    try {
+      ok = GroupMembership.hasModeratorRole(req.session.userId, p.groupId)
+    } catch (err) {
+      return res.status(422).send({ error: err.message ? err.message : err })
+    }
+
+    if (!ok) {
+      return res.status(403).send({ error: "No access" })
+    }
+
+    // process specified datasets
+    if (p.datasets.includes('members')) {
+      return exportMembers(p.groupId, req, res)
+    }
+
+    // got to the end and nothing output/exited, throw error
+    throw new Error("Unknown datasets specified: " + JSON.stringify(p.datasets))
+  }
+}
+
+/**
+ * Group members export by Group ID
+ */
+async function exportMembers(groupId, req, res) {
+  const users = await new Group({ id: groupId })
+                  .members()
+                  .fetch()
+
+  const results = []
+
+  // iterate over all group members
+  await Promise.all(users.map((u, idx) => {
+    // pluck core user data into results
+    results.push(u.pick([
+      'name', 'contact_email', 'contact_phone', 'location', 'avatar_url', 'tagline', 'bio',
+      'url', 'twitter_url', 'facebook_url', 'linkedin_url'
+    ]))
+
+    // return combined promise to load all dependent user data and
+    // assign final child query results back onto matching result objects upon completion
+    return Promise.all([
+
+      // affiliations
+      u.affiliations().fetch()
+        .then(affils => {
+          results[idx].affiliations = accumulatePivotCell(affils, renderAffiliation)
+        }),
+
+      // skills
+      u.skills().fetch()
+        .then(skills => {
+          results[idx].skills = accumulatePivotCell(skills, renderSkill)
+        }),
+
+      // skills to learn
+      u.skillsToLearn().fetch()
+        .then(skills => {
+          results[idx].skills_to_learn = accumulatePivotCell(skills, renderSkill)
+        }),
+
+      // Join request questions & answers
+      u.joinRequests().where({ group_id: groupId, status: JoinRequest.STATUS.Accepted }).fetch()
+        .then(jrs => Promise.all(jrs.map(jr =>
```

Comment:
hmm, shouldnt there only ever be one accepted join request? lets just sort by created_at DESC and choose the one most recent accepted one (its possible they could join then leave the group then request to join again and be accepted again)